MCPClient: disable capture of output for compressAIP client script

lower29 opened the pull request:
The task in #39 is to disable the capture of output specifically for the `compressAIP` function in `compressAIP.py` in the MCP client scripts. However, in looking at the code it became apparent that the best place to do this is actually in the `executeOrRunSubprocess.py` library in `archivematicaCommon`.

I've therefore added a `capture_output` parameter to the functions in `executeOrRunSubprocess.py`. This is set to True by default so existing behaviour is maintained. However the compressAIP script sets `capture_output` to `False`, thereby disabling the capture of stdout and stderr for this specific task.

I've also included a `py.test` unit test to demonstrate and prove the expected functionality.

@jhsimpson and @sevein I'd appreciate your feedback on this approach, in light of the work that is in progress or planned upstream.

Commit:
MCPClient: disable capture of output for compressAIP client script

This adds a capture_output parameter to the functions in
executeOrRunSubprocess.py. This is set to True by default so existing
behaviour is maintained. However the compressAIP script sets capture_output
to False, thereby disabling the capture of stdout and stderr for this
specific task.

Includes unit test to demonstrate expected functionality.

This fixes JiscRDSS issue #39.

sevein commented:
@lower29 #37 is now merged, it included a fix for the linting issue that breaks your build.

jrwdunham commented:
@lower29 FYI I've done some profiling and it seems that if you pass stdout/stderr to devnull in your `Popen` call, you'll get significant performance gains. 

To replicate, create a Python executable file called bigoutput containing this:

    #!/usr/bin/env python
    from __future__ import print_function
    SIZE = 10 ** 7
    print('.' * SIZE)

Then in iPython I timed both the "with-devnull" and "without-devnull" (`with_popen`) strategies:

    In [24]: def with_popen():
        ...:     p = subprocess.Popen(['./bigoutput'])
        ...:     p.communicate()

    In [25]: def with_devnull():
        ...:     with open(os.devnull, 'w') as devnull:
        ...:         p = subprocess.Popen(['./bigoutput'], stdout=devnull, stderr=devnull)
        ...:         p.communicate()

    In [26]: %time with_popen()
    CPU times: user 643 µs, sys: 5.8 ms, total: 6.44 ms
    Wall time: 338 ms

    In [27]: %time with_devnull()
    CPU times: user 589 µs, sys: 3.6 ms, total: 4.19 ms
    Wall time: 42.8 ms

I also tried passing various values for the `bufsize` arg to `Popen` and that had little effect in terms of improving the bare `Popen` strategy.

I'm going to be doing something similar to this in the https://github.com/artefactual/archivematica so hopefully we can coordinate efforts.

scollazo committed:
Disable MCPClient output by default

jrwdunham committed:
Pass stdout/stderr to devnull for no-capture Popen

jrwdunham committed:
Fix launchSubProcess tests

- Now that the `capture_output` kwarg defaults to `False`, the tests had to be
  modified to account for that.
- Also fixed some formatting/linter issues in the test file.

Commit:
Disable MCPClient output

lower29 commented:
Thanks to @jrwdunham and others for the work upstream on https://github.com/artefactual/archivematica/pull/805. I've rebased the `dev/jisc-issue-39` branch against latest `qa/jisc` and have cherry-picked the commits from the upstream PR. I had to make a change when merging https://github.com/artefactual/archivematica/pull/805/commits/cbafeaf3b1814b126eb3fd7ba9c080d417ffc888 because there was an existing change 335e942b37e4e9bf15b052a244cfdba8f1575d9f in this fork that caused a conflict.

@jhsimpson I think this is now ready for review in the context of the Jisc fork too - I realise that it's ahead of vanilla in that this came from 1.8.x, so we've got an impure mix of 1.6.x and other bits on this fork. No doubt this will work itself out in the end!

helenst commented on src/archivematicaCommon/lib/executeOrRunSubProcess.py:
```diff
@@ -28,7 +28,8 @@
 
-def launchSubProcess(command, stdIn="", printing=True, arguments=[], env_updates={}):
+def launchSubProcess(command, stdIn="", printing=True, arguments=[],
+                     env_updates={}, capture_output=False):
```
The docstring says "defaults to `True`" but this is defaulting to `False`. Elsewhere, this defaults to `True` so I think it should be that way here as well.

lower29 replied:
My original version was actually `capture_output=True` because that was the behaviour before this change, but @scollazo's patch https://github.com/JiscRDSS/archivematica/pull/42/commits/48a9d25421c4cbf6640d941612e6e45d9db47fcd (from upstream) changed this to `capture_output=False`. I think, going forward, that **not** capturing output by default is the correct way to go, so I think it should remain `capture_output=False`.

That said, I take your point about the docstring comment, so will update this to reflect the code (i.e. default is `False`).

helenst commented on src/archivematicaCommon/tests/test_execute_functions.py:
```diff
@@ -0,0 +1,18 @@
+def test_capture_output():
+    """Tests behaviour of capture_output when executing sub processes."""
+
+    # Test that stdout and stderr are not captured by default
```
Regarding the earlier comment about `True`/`False` defaults - if `capture_output` is `True` by default for `launchSubProcess` then these tests will need updating - one to test the `True` default, and one passing `False`.

lower29 replied:
As above, I think the correct approach is for the default value to be `False`, but I will add an additional test to confirm that ommitting the parameter results in **no output**, and that specifying `True` and `False` also produces the right result.

Commit:
MCPClient: fix docstring for capture_output and add test

The docstring for launchSubProcess() was incorrectly stating that the
default value for capture_output was True, when it's actually False.

I've also added a test to confirm that capture_output is handled correctly
when capture_output is not enabled, as previously we were only testing
default and True conditions.

lower29 commented:
@helenst @jrwdunham @jhsimpson I've responded to the comments in Helen's review and made the changes she requested.

I was hoping to be able to run the unit tests to confirm they were working correctly, but couldn't figure out the correct incantation to get them to run (I even talked to @sevein about it but still no luck). I'm fairly certain it's correct, but haven't been able to confirm.

@helenst if you get the chance can you take another look and hopefully approve? @jhsimpson if you know that Helen's not going to be able to do this in a reasonable timeframe can you or @jrwdunham review again instead? It would be good to get this merged before Christmas this year. Thanks 🙂

helenst commented on src/archivematicaCommon/tests/test_execute_functions.py:
```diff
@@ -0,0 +1,25 @@
+    # Test that stdout and stderr are not captured when `capture_output` is
+    # not enabled.
+    ret, std_out, std_err = execsub.launchSubProcess(
+        ['ls', '/tmp'], capture_output=True)
```
I was able to run the tests as follows:
```
COMPOSE_FILE=docker-compose.dev.yml:docker-compose.qa.yml docker-compose run --rm --no-deps --entrypoint py.test --workdir /src/ archivematica-dashboard archivematicaCommon/tests/test_execute_functions.py
```

(running a full test suite causes other failures and I just wanted to focus on this one).

I think this one should be `capture_output=False`. (to match the comment and also make the test pass). Otherwise looks good :)

Commit:
fixup - correct logic in test case call

lower29 commented:
Hi @helenst, thanks for the magic incantation for running the test and thanks for spotting the stupid copy-paste error. I've fixed it now so the test runs as expected.

Its a shame we can't get this tested without using the rdss-archivematica compose, but for now it does the job, and the end result is more important than how we got there.

jhsimpson commented:
This work has been released upstream in archivematica v1.7.1 - @sevein you suggested closing this PR and merging in the 1.7.1 changes instead, I agree.

sevein commented:
https://github.com/JiscRDSS/archivematica/pull/78 replaces this.
